flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/delegation/PlannerBase.scala

```diff
@@ -169,17 +171,53 @@ abstract class PlannerBase(
   override def getParser: Parser = {
     if (parser == null || getTableConfig.getSqlDialect != currentDialect) {
       dialectFactory = getDialectFactory
-      parser =
-        dialectFactory.create(new DefaultParserContext(catalogManager, plannerContext, executor))
+      parser = dialectFactory.create(
+        new DefaultCalciteContext(catalogManager, getOperationTreeBuilder, plannerContext))
     }
     parser
   }
 
+  private def getOperationTreeBuilder: OperationTreeBuilder = {
+    OperationTreeBuilderImpl.create(
+      tableConfig,
+      classLoader,
+      functionCatalog.asLookup(f => getParser.parseIdentifier(f)),
+      catalogManager.getDataTypeFactory,
+      (path: String) => getTableReferenceExpression(path),
+      (s: String, inputRowType: RowType, outputType) =>
+        getParser.parseSqlExpression(s, inputRowType, outputType),
+      isStreamingMode
+    )
+  }
```

Comment:
The creation of OperationTreeBuilderImpl is super complex, and the logic is duplicated. This is difficult to maintain logic consistency in the future. 

Besides, the extracted `OperationTreeBuilder` seems incomplete (e.g, join) and is never used. I would prefer to introduce the `OperationTreeBuilder` when we indeed need this, otherwise, it only increases the overhead of maintenance. What do you think?

Reply:
I do agree. We can add it in the future when we do need it. That does make everything simple.

Btw, the reason that `OperationTreeBuilder` miss join is the `JoinType` is not marked as `PublicEvolving `, so we can't expose it in `OperationTreeBuilder`.